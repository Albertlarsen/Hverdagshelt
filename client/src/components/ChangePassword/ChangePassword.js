import ReactDOM from "react-dom";
import * as React from "react";
import { Component } from "react-simplified";
import { Col, Grid, FormGroup, ControlLabel, FormControl, Button, Form } from "react-bootstrap";
import { UserService, MailService } from "../../services";
import jwt from "jsonwebtoken";
import { User } from "../../classTypes";
import css from './ChangePassword.css';

let userService = new UserService();
let mailService = new MailService();
const bcrypt = require('bcrypt-nodejs');

export class ChangePassword extends Component {
  constructor(props) {
    super(props);
    this.state = {
      user: User,
      oldPassword: '',
      newPassword: '',
      newPassword2: ''
    };
  }

  componentDidMount() {
    userService.getCurrentUser().then(newUser => {
      this.setState({
        user: newUser[0]
      });
    });
  }

  handleOldPassChange(e: Object) {
    this.setState({
      oldPassword: event.target.value
    });
  }

  handleNewPassChange(e: Object) {
    this.setState({
      newPassword: event.target.value
    });
  }

  handleNewPass2Change(e: Object) {
    this.setState({
      newPassword2: event.target.value
    });
  }
    getValidationStatePassword(){
        const length = this.state.newPassword.length;
        let decimal = /(?=^.{8,64}$)(?=.*\d)(?=.*[a-z])(?=.*[A-Z])(?!.*\s)[0-9a-zA-Z!., æøå/@<>"¤=#$%^&*()]*$/;
        if (this.state.newPassword.match(decimal)) return 'success';
        else if(length==0)return ;
        else return 'warning';
    }

    getValidationStatePassword2(){
        const password2Length = this.state.newPassword2.length;
        let decimal = /(?=^.{8,64}$)(?=.*\d)(?=.*[a-z])(?=.*[A-Z])(?!.*\s)[0-9a-zA-Z!., æøå/@<>"¤=#$%^&*()]*$/;
        if(password2Length==0) return;
        else{
            if(this.state.newPassword !== this.state.newPassword2||!(this.state.newPassword2.match(decimal))) return 'warning';
            else return 'success';
        }
    }

  submit() {
    if (this.state.newPassword == this.state.newPassword2) {
      bcrypt.compare(
        this.state.oldPassword,
        this.state.user.password,
        (err, res) => {
          if (res) {
            userService.resetPassword({
              newPassword: this.state.newPassword,
              email: this.state.email
            });
            alert('Passordet ditt er endret');
            console.log(this.state.user.mail);
            mailService.sendTextMail({
              to: this.state.user.mail,
              subject: 'Hverdagshelt - Endring av passord',
              text:
                'Hei ' +
                this.state.user.firstName +
                ', endringen av passordet ditt var vellykket. Om det ikke var deg som forespurte endringen, vennligst ta kontakt med oss'
            });
          } else {
            alert('Feil passord');
          }
        }
      );
    } else {
      alert('Passordene samsvarer ikke');
    }
  }

  render() {
    return (
      <div className="changePassword">
        <Form>
          <FormGroup controlId="formInlineToPassword">
            <ControlLabel>Gammelt passord</ControlLabel>{' '}
            <FormControl
              name="oldPassword"
              onChange={this.handleOldPassChange}
              type="password"
              placeholder="passord"
            />
          </FormGroup>{' '}
          <FormGroup controlId="formInlineToPassword" validationState={this.getValidationStatePassword()}>
            <ControlLabel>Nytt passord</ControlLabel>{' '}
            <FormControl
              name="newPassword"
              onChange={this.handleNewPassChange}
              type="password"
              placeholder="nytt passord"
            />
            <FormControl.Feedback/>
          </FormGroup>{' '}
          <FormGroup controlId="formInlineToPassword" validationState={this.getValidationStatePassword2()}>
            <ControlLabel>Gjenta nytt passord</ControlLabel>{' '}
            <FormControl
              name="newPassword2"
              onChange={this.handleNewPass2Change}
              type="password"
              placeholder="nytt passord"
            />
            <FormControl.Feedback/>
          </FormGroup>{' '}
          <div align="right">
            <Button onClick={this.submit} bsStyle="primary">
              Lagre nytt passord
            </Button>
          </div>
        </Form>
      </div>
    );
  }
}
