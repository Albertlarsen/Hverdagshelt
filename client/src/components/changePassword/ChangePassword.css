.changePassword{
    padding-top: 5em;
}