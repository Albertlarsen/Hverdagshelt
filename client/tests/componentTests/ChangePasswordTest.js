// @flow

import {shallow} from "enzyme/build";
import {ChangePassword} from "../../src/components/ChangePassword/ChangePassword";
import * as React from "react";

describe('Test for ChangePassword component', () => {
    const wrapper = shallow(<ChangePassword />);


    it('renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

});