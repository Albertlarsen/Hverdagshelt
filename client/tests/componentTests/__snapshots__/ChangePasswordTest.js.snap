// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for ChangePassword component renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ChangePassword />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <Form
        bsClass="form"
        componentClass="form"
        horizontal={false}
        inline={false}
      >
        <FormGroup
          bsClass="form-group"
          controlId="formInlineToPassword"
        >
          <ControlLabel
            bsClass="control-label"
            srOnly={false}
          >
            Gammelt passord
          </ControlLabel>
           
          <FormControl
            bsClass="form-control"
            componentClass="input"
            name="oldPassword"
            onChange={[Function]}
            placeholder="passord"
            type="password"
          />
        </FormGroup>
         
        <FormGroup
          bsClass="form-group"
          controlId="formInlineToPassword"
        >
          <ControlLabel
            bsClass="control-label"
            srOnly={false}
          >
            Nytt passord
          </ControlLabel>
           
          <FormControl
            bsClass="form-control"
            componentClass="input"
            name="newPassword"
            onChange={[Function]}
            placeholder="nytt passord"
            type="password"
          />
        </FormGroup>
         
        <FormGroup
          bsClass="form-group"
          controlId="formInlineToPassword"
        >
          <ControlLabel
            bsClass="control-label"
            srOnly={false}
          >
            Gjenta nytt passord
          </ControlLabel>
           
          <FormControl
            bsClass="form-control"
            componentClass="input"
            name="newPassword2"
            onChange={[Function]}
            placeholder="nytt passord"
            type="password"
          />
        </FormGroup>
         
        <div
          align="right"
        >
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="primary"
            disabled={false}
            onClick={[Function]}
          >
            Lagre nytt passord
          </Button>
        </div>
      </Form>,
      "className": "changePassword",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "bsClass": "form",
        "children": Array [
          <FormGroup
            bsClass="form-group"
            controlId="formInlineToPassword"
          >
            <ControlLabel
              bsClass="control-label"
              srOnly={false}
            >
              Gammelt passord
            </ControlLabel>
             
            <FormControl
              bsClass="form-control"
              componentClass="input"
              name="oldPassword"
              onChange={[Function]}
              placeholder="passord"
              type="password"
            />
          </FormGroup>,
          " ",
          <FormGroup
            bsClass="form-group"
            controlId="formInlineToPassword"
          >
            <ControlLabel
              bsClass="control-label"
              srOnly={false}
            >
              Nytt passord
            </ControlLabel>
             
            <FormControl
              bsClass="form-control"
              componentClass="input"
              name="newPassword"
              onChange={[Function]}
              placeholder="nytt passord"
              type="password"
            />
          </FormGroup>,
          " ",
          <FormGroup
            bsClass="form-group"
            controlId="formInlineToPassword"
          >
            <ControlLabel
              bsClass="control-label"
              srOnly={false}
            >
              Gjenta nytt passord
            </ControlLabel>
             
            <FormControl
              bsClass="form-control"
              componentClass="input"
              name="newPassword2"
              onChange={[Function]}
              placeholder="nytt passord"
              type="password"
            />
          </FormGroup>,
          " ",
          <div
            align="right"
          >
            <Button
              active={false}
              block={false}
              bsClass="btn"
              bsStyle="primary"
              disabled={false}
              onClick={[Function]}
            >
              Lagre nytt passord
            </Button>
          </div>,
        ],
        "componentClass": "form",
        "horizontal": false,
        "inline": false,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "bsClass": "form-group",
            "children": Array [
              <ControlLabel
                bsClass="control-label"
                srOnly={false}
              >
                Gammelt passord
              </ControlLabel>,
              " ",
              <FormControl
                bsClass="form-control"
                componentClass="input"
                name="oldPassword"
                onChange={[Function]}
                placeholder="passord"
                type="password"
              />,
            ],
            "controlId": "formInlineToPassword",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "bsClass": "control-label",
                "children": "Gammelt passord",
                "srOnly": false,
              },
              "ref": null,
              "rendered": "Gammelt passord",
              "type": [Function],
            },
            " ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "bsClass": "form-control",
                "componentClass": "input",
                "name": "oldPassword",
                "onChange": [Function],
                "placeholder": "passord",
                "type": "password",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        " ",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "bsClass": "form-group",
            "children": Array [
              <ControlLabel
                bsClass="control-label"
                srOnly={false}
              >
                Nytt passord
              </ControlLabel>,
              " ",
              <FormControl
                bsClass="form-control"
                componentClass="input"
                name="newPassword"
                onChange={[Function]}
                placeholder="nytt passord"
                type="password"
              />,
            ],
            "controlId": "formInlineToPassword",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "bsClass": "control-label",
                "children": "Nytt passord",
                "srOnly": false,
              },
              "ref": null,
              "rendered": "Nytt passord",
              "type": [Function],
            },
            " ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "bsClass": "form-control",
                "componentClass": "input",
                "name": "newPassword",
                "onChange": [Function],
                "placeholder": "nytt passord",
                "type": "password",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        " ",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "bsClass": "form-group",
            "children": Array [
              <ControlLabel
                bsClass="control-label"
                srOnly={false}
              >
                Gjenta nytt passord
              </ControlLabel>,
              " ",
              <FormControl
                bsClass="form-control"
                componentClass="input"
                name="newPassword2"
                onChange={[Function]}
                placeholder="nytt passord"
                type="password"
              />,
            ],
            "controlId": "formInlineToPassword",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "bsClass": "control-label",
                "children": "Gjenta nytt passord",
                "srOnly": false,
              },
              "ref": null,
              "rendered": "Gjenta nytt passord",
              "type": [Function],
            },
            " ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "bsClass": "form-control",
                "componentClass": "input",
                "name": "newPassword2",
                "onChange": [Function],
                "placeholder": "nytt passord",
                "type": "password",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        " ",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "align": "right",
            "children": <Button
              active={false}
              block={false}
              bsClass="btn"
              bsStyle="primary"
              disabled={false}
              onClick={[Function]}
            >
              Lagre nytt passord
            </Button>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "active": false,
              "block": false,
              "bsClass": "btn",
              "bsStyle": "primary",
              "children": "Lagre nytt passord",
              "disabled": false,
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Lagre nytt passord",
            "type": [Function],
          },
          "type": "div",
        },
      ],
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <Form
          bsClass="form"
          componentClass="form"
          horizontal={false}
          inline={false}
        >
          <FormGroup
            bsClass="form-group"
            controlId="formInlineToPassword"
          >
            <ControlLabel
              bsClass="control-label"
              srOnly={false}
            >
              Gammelt passord
            </ControlLabel>
             
            <FormControl
              bsClass="form-control"
              componentClass="input"
              name="oldPassword"
              onChange={[Function]}
              placeholder="passord"
              type="password"
            />
          </FormGroup>
           
          <FormGroup
            bsClass="form-group"
            controlId="formInlineToPassword"
          >
            <ControlLabel
              bsClass="control-label"
              srOnly={false}
            >
              Nytt passord
            </ControlLabel>
             
            <FormControl
              bsClass="form-control"
              componentClass="input"
              name="newPassword"
              onChange={[Function]}
              placeholder="nytt passord"
              type="password"
            />
          </FormGroup>
           
          <FormGroup
            bsClass="form-group"
            controlId="formInlineToPassword"
          >
            <ControlLabel
              bsClass="control-label"
              srOnly={false}
            >
              Gjenta nytt passord
            </ControlLabel>
             
            <FormControl
              bsClass="form-control"
              componentClass="input"
              name="newPassword2"
              onChange={[Function]}
              placeholder="nytt passord"
              type="password"
            />
          </FormGroup>
           
          <div
            align="right"
          >
            <Button
              active={false}
              block={false}
              bsClass="btn"
              bsStyle="primary"
              disabled={false}
              onClick={[Function]}
            >
              Lagre nytt passord
            </Button>
          </div>
        </Form>,
        "className": "changePassword",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "bsClass": "form",
          "children": Array [
            <FormGroup
              bsClass="form-group"
              controlId="formInlineToPassword"
            >
              <ControlLabel
                bsClass="control-label"
                srOnly={false}
              >
                Gammelt passord
              </ControlLabel>
               
              <FormControl
                bsClass="form-control"
                componentClass="input"
                name="oldPassword"
                onChange={[Function]}
                placeholder="passord"
                type="password"
              />
            </FormGroup>,
            " ",
            <FormGroup
              bsClass="form-group"
              controlId="formInlineToPassword"
            >
              <ControlLabel
                bsClass="control-label"
                srOnly={false}
              >
                Nytt passord
              </ControlLabel>
               
              <FormControl
                bsClass="form-control"
                componentClass="input"
                name="newPassword"
                onChange={[Function]}
                placeholder="nytt passord"
                type="password"
              />
            </FormGroup>,
            " ",
            <FormGroup
              bsClass="form-group"
              controlId="formInlineToPassword"
            >
              <ControlLabel
                bsClass="control-label"
                srOnly={false}
              >
                Gjenta nytt passord
              </ControlLabel>
               
              <FormControl
                bsClass="form-control"
                componentClass="input"
                name="newPassword2"
                onChange={[Function]}
                placeholder="nytt passord"
                type="password"
              />
            </FormGroup>,
            " ",
            <div
              align="right"
            >
              <Button
                active={false}
                block={false}
                bsClass="btn"
                bsStyle="primary"
                disabled={false}
                onClick={[Function]}
              >
                Lagre nytt passord
              </Button>
            </div>,
          ],
          "componentClass": "form",
          "horizontal": false,
          "inline": false,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "bsClass": "form-group",
              "children": Array [
                <ControlLabel
                  bsClass="control-label"
                  srOnly={false}
                >
                  Gammelt passord
                </ControlLabel>,
                " ",
                <FormControl
                  bsClass="form-control"
                  componentClass="input"
                  name="oldPassword"
                  onChange={[Function]}
                  placeholder="passord"
                  type="password"
                />,
              ],
              "controlId": "formInlineToPassword",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "bsClass": "control-label",
                  "children": "Gammelt passord",
                  "srOnly": false,
                },
                "ref": null,
                "rendered": "Gammelt passord",
                "type": [Function],
              },
              " ",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "bsClass": "form-control",
                  "componentClass": "input",
                  "name": "oldPassword",
                  "onChange": [Function],
                  "placeholder": "passord",
                  "type": "password",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          " ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "bsClass": "form-group",
              "children": Array [
                <ControlLabel
                  bsClass="control-label"
                  srOnly={false}
                >
                  Nytt passord
                </ControlLabel>,
                " ",
                <FormControl
                  bsClass="form-control"
                  componentClass="input"
                  name="newPassword"
                  onChange={[Function]}
                  placeholder="nytt passord"
                  type="password"
                />,
              ],
              "controlId": "formInlineToPassword",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "bsClass": "control-label",
                  "children": "Nytt passord",
                  "srOnly": false,
                },
                "ref": null,
                "rendered": "Nytt passord",
                "type": [Function],
              },
              " ",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "bsClass": "form-control",
                  "componentClass": "input",
                  "name": "newPassword",
                  "onChange": [Function],
                  "placeholder": "nytt passord",
                  "type": "password",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          " ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "bsClass": "form-group",
              "children": Array [
                <ControlLabel
                  bsClass="control-label"
                  srOnly={false}
                >
                  Gjenta nytt passord
                </ControlLabel>,
                " ",
                <FormControl
                  bsClass="form-control"
                  componentClass="input"
                  name="newPassword2"
                  onChange={[Function]}
                  placeholder="nytt passord"
                  type="password"
                />,
              ],
              "controlId": "formInlineToPassword",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "bsClass": "control-label",
                  "children": "Gjenta nytt passord",
                  "srOnly": false,
                },
                "ref": null,
                "rendered": "Gjenta nytt passord",
                "type": [Function],
              },
              " ",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "bsClass": "form-control",
                  "componentClass": "input",
                  "name": "newPassword2",
                  "onChange": [Function],
                  "placeholder": "nytt passord",
                  "type": "password",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          " ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "align": "right",
              "children": <Button
                active={false}
                block={false}
                bsClass="btn"
                bsStyle="primary"
                disabled={false}
                onClick={[Function]}
              >
                Lagre nytt passord
              </Button>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "active": false,
                "block": false,
                "bsClass": "btn",
                "bsStyle": "primary",
                "children": "Lagre nytt passord",
                "disabled": false,
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Lagre nytt passord",
              "type": [Function],
            },
            "type": "div",
          },
        ],
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
